stablehlo/reference/Ops.cpp

```diff
@@ -76,6 +76,97 @@ Tensor eval(CosineOp op, const Tensor &operand) {
+      op.getDotDimensionNumbers().getLhsBatchingDimensions();
+  ArrayRef<int64_t> rhsBatchingDims =
+      op.getDotDimensionNumbers().getRhsBatchingDimensions();
+  ArrayRef<int64_t> lhsContractingDims =
+      op.getDotDimensionNumbers().getLhsContractingDimensions();
+  ArrayRef<int64_t> rhsContractingDims =
+      op.getDotDimensionNumbers().getRhsContractingDimensions();
+  std::vector<int64_t> lhsNonContractingDims;
+  for (int64_t i = 0; i < lhs.getType().getRank(); ++i)
+    if (std::find(lhsBatchingDims.begin(), lhsBatchingDims.end(), i) ==
+            lhsBatchingDims.end() &&
+        std::find(lhsContractingDims.begin(), lhsContractingDims.end(), i) ==
+            lhsContractingDims.end())
+      lhsNonContractingDims.push_back(i);
+  std::vector<int64_t> rhsNonContractingDims;
+  for (int64_t i = 0; i < rhs.getType().getRank(); ++i)
+    if (std::find(rhsBatchingDims.begin(), rhsBatchingDims.end(), i) ==
+            rhsBatchingDims.end() &&
+        std::find(rhsContractingDims.begin(), rhsContractingDims.end(), i) ==
+            rhsContractingDims.end())
+      rhsNonContractingDims.push_back(i);
+  std::vector<int64_t> contractingDimSizes;
+  for (uint64_t i = 0; i < lhsContractingDims.size(); ++i)
+    contractingDimSizes.push_back(lhsShape[lhsContractingDims[i]]);
+  auto totalContractingSize = 1;
+  for (uint64_t i = 0; i < contractingDimSizes.size(); ++i)
+    totalContractingSize *= contractingDimSizes[i];
+  for (auto resIt = result.index_begin(); resIt != result.index_end();
+       ++resIt) {
+    Type resElTy =
+        op->getResultTypes().front().dyn_cast<ShapedType>().getElementType();
+    Element sum;
+    if (isSupportedSignedIntegerType(resElTy)) {
+      sum = Element(resElTy, APInt(resElTy.getIntOrFloatBitWidth(), 0,
+                                   /*isSigned=*/true));
+    } else if (isSupportedUnsignedIntegerType(resElTy)) {
+      sum = Element(resElTy, APInt(resElTy.getIntOrFloatBitWidth(), 0,
+                                   /*isSigned=*/false));
+    } else if (isSupportedBooleanType(resElTy)) {
+      sum = Element(resElTy, false);
+    } else if (isSupportedFloatType(resElTy)) {
+      APFloat val((double)0.0);
+      bool roundingErr;
+      val.convert(resElTy.cast<FloatType>().getFloatSemantics(),
+                  APFloat::rmNearestTiesToEven, &roundingErr);
+      sum = Element(resElTy, val);
+    } else if (isSupportedComplexType(resElTy)) {
+      APFloat real((double)0.0);
+      APFloat imag((double)0.0);
+      FloatType flType =
+          resElTy.cast<ComplexType>().getElementType().cast<FloatType>();
+      bool roundingErr;
+      real.convert(flType.getFloatSemantics(), APFloat::rmNearestTiesToEven,
+                   &roundingErr);
+      imag.convert(flType.getFloatSemantics(), APFloat::rmNearestTiesToEven,
+                   &roundingErr);
+      sum = Element(resElTy, std::complex<APFloat>(real, imag));
+    }
+    std::vector<int64_t> lhsIdx(lhs.getType().getRank());
```

Comment:
Can you add some comment here (may be using a simple example) to explain what are we doing below?

Reply:
Something like for batching and non-batching-non-constracting dimensions the indices of result, lhs and rhs are the same.